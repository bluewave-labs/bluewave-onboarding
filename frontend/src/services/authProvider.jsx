import React, { useEffect, useContext, useReducer, useState } from 'react';
import { apiClient } from './apiClient';

const AuthContext = React.createContext();

export const useAuth = () => {
    return useContext(AuthContext);
};

// HTML entity decoder function
const decodeHtmlEntities = (str) => {
    if (!str) return str;
    const parser = new DOMParser();
    const dom = parser.parseFromString(
        `<!doctype html><body>${str}`, 'text/html'
    );
    return dom.body.textContent;
};

// Function to handle user data processing
const processUserData = (userData) => {
    return {
        ...userData,
        picture: userData.picture ? decodeHtmlEntities(userData.picture) : null
    };
};

const authReducer = (state, action) => {
    switch (action.type) {
        case 'LOGIN':
            return { ...state, isLoggedIn: true };
        case 'LOGOUT':
            return { ...state, isLoggedIn: false, userInfo: null };
        case 'SET_USER_INFO': {
            const processedData = processUserData(action.payload);
            localStorage.setItem('userInfo', JSON.stringify(processedData));
            return { ...state, userInfo: processedData };
        }
        case 'LOGIN_AND_SET_USER_INFO': {
            const processedData = processUserData(action.payload);
            localStorage.setItem('userInfo', JSON.stringify(processedData));
            return { isLoggedIn: true, userInfo: processedData };
        }
        case 'UPDATE_AND_SET_UPDATED_USER_INFO': {
            const updatedUserInfo = processUserData({
                ...state.userInfo,
                ...action.payload,
            });
            localStorage.setItem('userInfo', JSON.stringify(updatedUserInfo));
            return {isLoggedIn: true, userInfo: updatedUserInfo };
        }
        default:
            return state;
    }
};

export const AuthProvider = ({ children }) => {
    // Decoding any existing stored data on initial load
    const getInitialState = () => {
        const storedData = localStorage.getItem('userInfo');
        if (storedData) {
            try {
                const parsedData = JSON.parse(storedData);
                return { 
                    isLoggedIn: false, 
                    userInfo: processUserData(parsedData)
                };
            } catch (e) {
                console.error('Error parsing stored user data:', e);
                return { isLoggedIn: false, userInfo: null };
            }
        }
        return { isLoggedIn: false, userInfo: null };
    };

    const [state, dispatch] = useReducer(authReducer, getInitialState());
    const [isFetching, setIsFetching] = useState(true);

    useEffect(() => {
        const fetchUser = async () => {
            try {
                const authToken = localStorage.getItem('authToken');
                if (!authToken) {
                    dispatch({ type: 'LOGOUT' });
                    setIsFetching(false);
                    return;
                }
                const response = await apiClient.get('/users/current-user');
                if (response.status === 200 && response.data.user) {
                    const { id, name, surname, email, role, picture } = response.data.user;
                    const payload = { id, name, surname, email, role, picture };
                    dispatch({ type: 'LOGIN_AND_SET_USER_INFO', payload });
                } else {
                    dispatch({ type: 'LOGOUT' });
                }
            } catch (error) {
                localStorage.removeItem('authToken');
                localStorage.removeItem('userInfo');
                dispatch({ type: 'LOGOUT' });
            } finally {
                setIsFetching(false);
            }
        };
        fetchUser();
    }, []);

    const loginAuth = (userInfo) => {
        dispatch({ type: 'LOGIN_AND_SET_USER_INFO', payload: userInfo });
    };

    const updateProfile = (userInfo) => {
        dispatch({ type: 'UPDATE_AND_SET_UPDATED_USER_INFO', payload: userInfo });
    };

    const logoutAuth = () => {
        dispatch({ type: 'LOGOUT' });
    };

    return (
        <AuthContext.Provider value={{ 
            isLoggedIn: state.isLoggedIn, 
            loginAuth, 
            logoutAuth, 
            updateProfile, 
            userInfo: state.userInfo, 
            isFetching
        }}>
            {children}
        </AuthContext.Provider>
    );
};