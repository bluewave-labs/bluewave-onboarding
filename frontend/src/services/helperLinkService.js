import { apiClient } from "./apiClient";

export const getHelpers = async () => {
  try {
    const response = await apiClient.get(`/helper-link/get_helpers`);
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Get helper link error:", error.response);
    throw error;
  }
};
export const getHelperById = async (id) => {
  try {
    const response = await apiClient.get(`/helper-link/get_helper/${id}`);
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Get helper link error:", error.response);
    throw error;
  }
};
const validateHelper = (helper, links) => {
  if (!helper?.title) {
    throw new Error("Header is required");
  }
  if (!helper?.headerBackgroundColor) {
    throw new Error("Header background color is required");
  }
  if (!helper?.linkFontColor) {
    throw new Error("Link font color is required");
  }
  if (!helper?.iconColor) {
    throw new Error("Icon color is required");
  }
  if (links.some((it) => !it?.title || !it?.url)) {
    throw new Error("Title and URL are required");
  }
};

export const createHelper = async (helper, links) => {
  try {
    validateHelper(helper, links);
    const response = await apiClient.post(`/helper-link/add_helper`, {
      title: helper.title.trim(),
      headerBackgroundColor: helper.headerBackgroundColor,
      linkFontColor: helper.linkFontColor,
      iconColor: helper.iconColor,
      userId: helper.userId,
      links,
    });
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Create helper link error:", error.message);
    throw error;
  }
};

export const updateHelper = async (helper, links) => {
  try {
    validateHelper(helper, links);
    const response = await apiClient.put(
      `/helper-link/edit_helper/${helper.id}`,
      {
        title: helper.title,
        headerBackgroundColor: helper.headerBackgroundColor,
        linkFontColor: helper.linkFontColor,
        iconColor: helper.iconColor,
        userId: helper.userId,
        links,
      }
    );
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Update helper link error:", error.response);
    throw error;
  }
};

export const deleteHelper = async (id) => {
  try {
    const response = await apiClient.delete(`/helper-link/delete_helper/${id}`);
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Delete helper link error:", error.response);
    throw error;
  }
};
