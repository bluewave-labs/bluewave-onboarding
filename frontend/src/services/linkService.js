import { apiClient } from "./apiClient";
const RELATIVE_URL_REGEX = /^([./?#]|[a-zA-Z0-9_-]+\/?)+$/;

const isValidUrl = (url) => {
  try {
    new URL(url);
    return true;
  } catch (e) {
    return RELATIVE_URL_REGEX.test(url);
  }
};

export const createLink = async (link) => {
  try {
    if (!link.title?.trim() || !link.url?.trim()) {
      throw new Error("Title and URL are required");
    }
    if (!isValidUrl(link.url)) {
      throw new Error("Invalid URL format");
    }
    const response = await apiClient.post(`/link/add_link`, {
      title: link.title,
      url: link.url,
      order: link.order,
      target: link.target ? "_blank" : "_self",
      helperId: link.helperId,
    });
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Create Link error:", error.response);
    throw error;
  }
};

export const getLinks = async (helperId) => {
  try {
    if (!helperId) {
      throw new Error("Helper ID is required");
    }
    const response = await apiClient.get(
      `/link/get_links?helperId=${helperId}`
    );
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Get Link error:", error.response);
    throw error;
  }
};

export const getLinkById = async (id) => {
  try {
    if (!id) {
      throw new Error("Link ID is required");
    }
    const response = await apiClient.get(`/link/get_link/${id}`);
    return response.data;
  } catch (error) {
    if (error.message.includes("Link ID is required")) {
      throw error;
    }
    return false;
  }
};

export const updateLink = async (link) => {
  try {
    if (!link.title?.trim() || !link.url?.trim()) {
      throw new Error("Title and URL are required");
    }
    if (!isValidUrl(link.url)) {
      throw new Error("Invalid URL format");
    }
    if (typeof link.order !== "number") {
      throw new Error("Order must be a number");
    }
    link.title = link.title.replace(/[<>]/g, "");
    const response = await apiClient.put(`/link/edit_link/${link.id}`, {
      title: link.title,
      url: link.url,
      order: link.order,
      target: link.target ? "_blank" : "_self",
      helperId: link.helperId,
    });
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Update Link error:", error.response);
    throw error;
  }
};

export const deleteLink = async (id) => {
  try {
    const response = await apiClient.delete(`/link/delete_link/${id}`);
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Delete Link error:", error.response);
    throw error;
  }
};
