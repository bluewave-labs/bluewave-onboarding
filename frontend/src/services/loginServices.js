import { apiClient } from "./apiClient";
import axios from "axios";
import Cookies from "js-cookie";

const authClient = axios.create({
  ...apiClient.defaults,
  baseURL: `${apiClient.defaults.baseURL}auth`,
});

// Function to handle login
export const login = async (values) => {
  try {
    const response = await authClient.post("/login", values);
    const token = response.data.token;
    localStorage.setItem("authToken", token);

    return response.data;
  } catch (error) {
    console.error("Login error:", error.response);
    throw error;
  }
};

// Function to handle logout
export const logout = async () => {
  try {
    await apiClient.post("auth/logout");
    localStorage.removeItem("authToken");
    return true;
  } catch (error) {
    console.error("Logout error:", error.response);
    localStorage.removeItem("authToken");
    throw error;
  }
};

// Function to handle sign up
export const signUp = async (userData) => {
  try {
    const response = await authClient.post("/register", userData);
    const token = response.data.token;
    localStorage.setItem("authToken", token);
    return response.data;
  } catch (error) {
    console.error("Sign up error:", error.response);
    throw error;
  }
};

// Function to handle sign up
export const forgotPassword = async (userData) => {
  try {
    const response = await authClient.post("/forget-password", userData);
    return response.data;
  } catch (error) {
    console.error(error.response);
    throw error;
  }
};

export const resetPassword = async (userData) => {
  try {
    const response = await authClient.post("/reset-password", userData);
    return response.data;
  } catch (error) {
    console.error(error.response);
    throw error;
  }
};

export const getCurrentUser = async () => {
  try {
    const response = await apiClient.get("users/current-user");
    const user = response.data.user;
    const fullName = user.surname ? user.name + " " + user.surname : user.name;

    Cookies.set("fullName", fullName);
    Cookies.set("role", user.role);

    return user;
  } catch (error) {
    console.error("Get user error:", error.response);
    return { fullName: "John Doe", role: "visitor" };
  }
};

export const getHasUsers = async () => {
  try {
    const response = await apiClient.get("users/has-users");
    return response.data;
  } catch (err) {
    throw err;
  }
};

export default authClient;
