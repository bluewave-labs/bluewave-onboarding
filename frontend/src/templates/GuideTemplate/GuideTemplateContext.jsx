import React, { createContext, useContext, useState } from "react";

const GuideTemplateContext = createContext();

export const useDialog = () => {
  const context = useContext(GuideTemplateContext);
  if (context === undefined) {
    throw new Error("useDialog must be used within a GuideTemplateProvider");
  }
  return context;
};

export const GuideTemplateProvider = ({ children }) => {
  const [isOpen, setIsOpen] = useState(false);

  const openDialog = () => setIsOpen(true);
  const closeDialog = () => setIsOpen(false);

  return (
    <GuideTemplateContext.Provider value={{ isOpen, openDialog, closeDialog }}>
      {children}
    </GuideTemplateContext.Provider>
  );
};
