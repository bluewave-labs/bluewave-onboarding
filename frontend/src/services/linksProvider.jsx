import PropTypes from "prop-types";
import { createContext, useMemo, useState } from "react";
import { getLinks } from "./linkService";

export const HelperLinkContext = createContext({
  helper: {},
  setHelper: (helper) => {},
  links: [],
  setLinks: (links) => {},
  showSettings: false,
  setShowSettings: (showSettings) => {},
  toggleSettings: (e, link) => {},
  itemToDelete: null,
  setItemToDelete: (link) => {},
  isPopupOpen: false,
  setIsPopupOpen: (openPopup) => {},
  deletedLinks: [],
  setDeletedLinks: (deletedLinks) => {},
  linkToEdit: null,
  setLinkToEdit: (link) => {},
  renderLinks: () => {},
  helperToEdit: null,
  setHelperToEdit: (helperId) => {},
});

const HelperLinkProvider = ({ children }) => {
  const [helper, setHelper] = useState({});
  const [links, setLinks] = useState([]);
  const [showSettings, setShowSettings] = useState(false);
  const [itemToDelete, setItemToDelete] = useState(null);
  const [isPopupOpen, setIsPopupOpen] = useState(false);
  const [deletedLinks, setDeletedLinks] = useState([]);
  const [linkToEdit, setLinkToEdit] = useState(null);
  const [helperToEdit, setHelperToEdit] = useState(null);

  const toggleSettings = (e, link = null) => {
    if (e.target.closest("#delete") || e.target.closest("#drag")) return;

    if (!showSettings && link) {
      setLinkToEdit(link);
    }
    setShowSettings(!showSettings); // Toggle the settings visibility
  };

  const renderLinks = async () => {
    if (helper.id) {
      const data = await getLinks(helper.id);
      setLinks(
        data
          .map((it) => ({ ...it, x: 0, y: 0 }))
          .sort((a, b) => a.order - b.order)
      );
    } else {
      renderLinks();
    }
  };

  const value = useMemo(
    () => ({
      helper,
      setHelper,
      links,
      setLinks,
      showSettings,
      setShowSettings,
      toggleSettings,
      itemToDelete,
      setItemToDelete,
      isPopupOpen,
      setIsPopupOpen,
      deletedLinks,
      setDeletedLinks,
      linkToEdit,
      setLinkToEdit,
      renderLinks,
      helperToEdit,
      setHelperToEdit,
    }),
    [helper, links, showSettings, isPopupOpen, linkToEdit]
  );
  return (
    <HelperLinkContext.Provider value={value}>
      {children}
    </HelperLinkContext.Provider>
  );
};

HelperLinkProvider.propTypes = {
  children: PropTypes.node,
};

export default HelperLinkProvider;
