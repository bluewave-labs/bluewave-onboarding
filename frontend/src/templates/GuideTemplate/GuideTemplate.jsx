import CloseOutlinedIcon from "@mui/icons-material/CloseOutlined";
import { Dialog } from "@mui/material";
import classNames from "classnames";
import PropTypes from "prop-types";
import { React } from "react";
import Button from "../../components/Button/Button";
import styles from "./GuideTemplate.module.scss";
import { useDialog } from "./GuideTemplateContext";

const GuideTemplate = ({
  title = "",
  handleButtonClick = () => null,
  activeButton = 0,
  leftContent = () => null,
  rightContent = () => null,
  leftAppearance = () => null,
  onSave = () => null,
}) => {
  const { isOpen, closeDialog } = useDialog();
  const buttons = ["Content", "Appearance"];

  return (
    <Dialog
      open={isOpen}
      onClose={closeDialog}
      maxWidth='lg'
      PaperProps={{ style: { position: "static" } }}
    >
      <div className={styles.container}>
        <div className={styles.popup}>
          <div className={styles.header}>
            <span style={{ marginLeft: "5px" }}>{title}</span>
            <CloseOutlinedIcon
              style={{
                color: "#98A2B3",
                fontSize: "20px",
                cursor: "pointer",
              }}
              onClick={closeDialog}
            />
          </div>
          <div className={styles.content}>
            {/* Content and Appereance buttons */}
            <div className={styles.buttons}>
              {buttons.map((buttonName, index) => (
                <button
                  key={index}
                  className={classNames(styles.menuButton, {
                    [styles.active]: activeButton === index,
                  })}
                  onClick={() => handleButtonClick(index)}
                >
                  {buttonName}
                </button>
              ))}
            </div>
            <div className={styles.leftRightContent}>
              {activeButton === 1 ? leftAppearance() : leftContent()}
              {rightContent()}
            </div>
            <div className={styles.optionButtons}>
              <Button
                text='Cancel'
                buttonType='secondary-grey'
                onClick={() => {
                  closeDialog();
                }}
              />
              <Button text='Save' onClick={onSave} />
            </div>
          </div>
        </div>
      </div>
    </Dialog>
  );
};

GuideTemplate.propTypes = {
  title: PropTypes.string,
  handleButtonClick: PropTypes.func,
  activeButton: PropTypes.number,
  leftContent: PropTypes.func,
  rightContent: PropTypes.func,
  leftAppearance: PropTypes.func,
  onSave: PropTypes.func,
};

export default GuideTemplate;
