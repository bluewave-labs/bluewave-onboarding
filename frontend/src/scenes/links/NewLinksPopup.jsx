import PropTypes from "prop-types";
import React, { useContext, useEffect, useState } from "react";
import Settings from "../../components/Links/Settings/Settings";
import Preview from "../../products/LinkPreview";
import {
  createHelper,
  getHelperById,
  updateHelper,
} from "../../services/helperLinkService";
import { deleteLink, getLinkById } from "../../services/linkService";
import { HelperLinkContext } from "../../services/linksProvider";
import GuideTemplate from "../../templates/GuideTemplate/GuideTemplate";
import { useDialog } from "../../templates/GuideTemplate/GuideTemplateContext";
import { emitToastError } from "../../utils/guideHelper";
import toastEmitter, { TOAST_EMITTER_KEY } from "../../utils/toastEmitter";
import styles from "./LinkPage.module.scss";
import LinkAppearance from "./LinkPageComponents/LinkAppearance";
import LinkContent from "./LinkPageComponents/LinkContent";

const NewLinksPopup = ({
  autoOpen = false,
  isEdit,
  itemId,
  setItemsUpdated,
}) => {
  const [activeBtn, setActiveBtn] = useState(0);
  const {
    showSettings,
    helper,
    setHelper,
    links,
    deletedLinks,
    setLinks,
    helperToEdit,
    setHelperToEdit,
  } = useContext(HelperLinkContext);

  const DEFAULT_VALUES = {
    title: "",
    headerBackgroundColor: "#F8F9F8",
    linkFontColor: "#344054",
    iconColor: "#7F56D9",
  }
  const { openDialog, closeDialog, isOpen } = useDialog();
  const fetchHelperData = async () => {
    try {
      const { links, ...data } = await getHelperById(itemId);
      setHelper(data);
      setLinks(links.sort((a, b) => a.order - b.order));
      setHelperToEdit(itemId);
    }
    catch (error) {
      emitToastError(buildToastError(error));
      closeDialog();
    }
  }

  useEffect(() => {
    if (autoOpen) {
      openDialog();
    }
  }, [autoOpen, openDialog]);

  useEffect(() => {
    if (isEdit) {
      fetchHelperData();
    } else {
      setHelper(DEFAULT_VALUES);
      setLinks([]);
    }
    if (!isOpen) {
      setLinks([]);
      setHelper({});
      setHelperToEdit(null);
    }
  }, [openDialog, isOpen]);

  const buildToastError = (msg) =>
    msg.response
      ? msg
      : {
        response: { data: { errors: [{ msg }] } },
      };

  const handleLinks = async (item) => {
    const { id, ...link } = item;
    if (!link?.title.trim() || !link?.url.trim()) {
      emitToastError(buildToastError("Title and URL are required"));
      return null;
    }
    try {
      const exists = await getLinkById(id);
      if (exists?.id) return { ...link, id };
      return { ...link };
    } catch (err) {
      emitToastError(err);
      return null;
    }
  };

  const handleSaveHelper = async () => {
    let newHelper;
    const formattedLinks = await Promise.all(links.map(handleLinks));
    try {
      if (formattedLinks.some((it) => !it)) {
        return null;
      }
      newHelper = await (helperToEdit
        ? updateHelper(helper, formattedLinks)
        : createHelper(helper, formattedLinks));
      setHelper(newHelper);
      setItemsUpdated((prevState) => !prevState);
      closeDialog();
    } catch (err) {
      emitToastError(buildToastError(err));
      return null;
    }
    if (helperToEdit && deletedLinks.length) {
      await Promise.all(
        deletedLinks.map(async (it) => {
          try {
            return await deleteLink({ ...it, helperId: helperToEdit });
          } catch (err) {
            emitToastError(err);
            return null;
          }
        })
      );
    }
    if (newHelper) {
      const toastMessage = helper.id
        ? "You edited this Helper Link"
        : "New Helper Link saved";
      toastEmitter.emit(TOAST_EMITTER_KEY, toastMessage);
      setShowNewLinksPopup(false);
      setHelper({});
      setLinks([]);
      setHelperToEdit(null);
    }
  };

  const rightContent = () => <Preview />;
  const leftContent = () => (
    <>
      {showSettings && <Settings />}
      <LinkContent />
    </>
  );
  const leftAppearance = () => <LinkAppearance />;

  return (
    <div className={styles.new__container}>
      <GuideTemplate
        title={isEdit ? "Edit Helper Link" : "New Helper Link"}
        activeButton={activeBtn}
        handleButtonClick={setActiveBtn}
        rightContent={rightContent}
        leftContent={leftContent}
        leftAppearance={leftAppearance}
        onSave={handleSaveHelper}
      />
    </div>
  );
};

NewLinksPopup.propTypes = {
  autoOpen: PropTypes.bool,
  isEdit: PropTypes.bool.isRequired,
  itemId: PropTypes.number,
  setItemsUpdated: PropTypes.func.isRequired,
};

export default NewLinksPopup;
