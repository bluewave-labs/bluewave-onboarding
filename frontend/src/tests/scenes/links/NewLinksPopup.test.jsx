import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { act } from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { describe, expect, it, vi } from "vitest";
import NewLinksPopup from "../../../scenes/links/NewLinksPopup";
import { AuthProvider } from "../../../services/authProvider";
import HelperLinkProvider from "../../../services/linksProvider";
import { GuideTemplateProvider } from "../../../templates/GuideTemplate/GuideTemplateContext";
import * as loginServices from "../../../services/loginServices";

vi.mock("../../../services/loginServices");
vi.mock("../../../services/helperLinkService");
vi.mock("../../../services/linkService");

const logIn = async () => {
  vi.spyOn(loginServices, "signUp").mockResolvedValueOnce({
    token: "mockAuthToken",
  });

  await act(async () => {
    await loginServices.signUp({
      name: "test",
      surname: "user",
      email: "test_user@example.com",
      password: "password123",
    });
    localStorage.setItem("authToken", "mockAuthToken");
  });
};

const renderPopup = async () => {
  await logIn();
  return render(
    <Router>
      <AuthProvider>
        <GuideTemplateProvider>
          <HelperLinkProvider>
            <NewLinksPopup
              autoOpen={true}
              currentHelper={{
                title: "",
                headerBackgroundColor: "#F8F9F8",
                linkFontColor: "#344054",
                iconColor: "#7F56D9",
              }}
            />
          </HelperLinkProvider>
        </GuideTemplateProvider>
      </AuthProvider>
    </Router>
  );
};

const addNewLink = async (title, url, target = true) => {
  const addNewLinkBtn = await screen.findByText("+ Add new link");
  await act(async () => {
    await userEvent.click(addNewLinkBtn);
  });
  const settings = await screen.findByTestId("settings-form");
  expect(settings).to.exist;
  const inputs = settings.querySelectorAll("input");
  await act(async () => {
    await userEvent.type(inputs[1], title);
    await userEvent.type(inputs[2], url);
    if (target === false) {
      await userEvent.click(inputs[3]);
    }
    await userEvent.click(screen.getByTestId("close"));
  });
};

const openAppearance = async () => {
  const appearanceBtn = await screen.findByRole("button", {
    name: "Appearance",
  });
  await act(async () => {
    await userEvent.click(appearanceBtn);
  });
};

describe("Test Helper Link popup", () => {
  it("Tests if the popup is rendered correctly", async () => {
    const { asFragment } = await renderPopup();
    await waitFor(() => {
      screen.debug(); // Log the rendered DOM
      expect(asFragment()).to.matchSnapshot();
    });
  });
  describe("Test adding a new link", () => {
    it("should open the settings drawer when add new link is clicked", async () => {
      await renderPopup();
      const addNewLinkBtn = await screen.findByText("+ Add new link");
      await act(async () => {
        await userEvent.click(addNewLinkBtn);
      });
      const settings = await screen.findByTestId("settings-form");
      expect(settings).to.exist;
      const inputs = settings.querySelectorAll("input");
      expect(inputs).toHaveLength(4);
      expect(inputs[0]).toHaveProperty("name", "id");
      expect(inputs[1]).toHaveProperty("name", "title");
      expect(inputs[2]).toHaveProperty("name", "url");
      expect(inputs[3]).toHaveProperty("name", "target");
    });
    it("should add the new link to links list and to the preview when a new link is created", async () => {
      await renderPopup();
      await addNewLink("link 1", "http://link1.com.br");
      const links = await screen.findAllByText("link 1");
      expect(links).toHaveLength(2);
      expect(links[1]).toHaveProperty("href", "http://link1.com.br/");
      expect(links[1]).toHaveProperty("target", "_blank");
    });
    it("should add the new link to links list and to the preview with the correct target when a new link is created", async () => {
      await renderPopup();
      await addNewLink("link 1", "http://link1.com.br");
      let links = await screen.findAllByText("link 1");
      expect(links).toHaveLength(2);
      expect(links[1]).toHaveProperty("href", "http://link1.com.br/");
      expect(links[1]).toHaveProperty("target", "_blank");

      await addNewLink("link 2", "http://link2.com.br", false);
      links = await screen.findAllByText("link 2");
      expect(links).toHaveLength(2);
      expect(links[1]).toHaveProperty("href", "http://link2.com.br/");
      expect(links[1]).toHaveProperty("target", "_self");
    });
    it.skip("should change the order of the links on preview after drag", async () => {
      await renderPopup();
      await addNewLink("link 1", "http://link1.com.br");
      await addNewLink("link 2", "http://link2.com.br");
      await addNewLink("link 3", "http://link3.com.br");
      const links = await screen.findAllByText(/link \d/, { selector: "span" });
      const linksContainer = screen.getByTestId("cards");
      const listItems = linksContainer.querySelectorAll("li");
      expect(links).toHaveLength(3);
      expect(links.map((it) => it.innerHTML)).toEqual([
        "link 1",
        "link 2",
        "link 3",
      ]);

      await act(async () => {
        await userEvent.pointer([
          { keys: "[MouseLeft>]", target: listItems[0] },
          { keys: "[MouseLeft]", target: listItems[2] },
          { keys: "[/MouseLeft]" },
        ]);
      });

      await waitFor(async () => {
        const updatedLinks = await screen.findAllByText(/link \d/, {
          selector: "span",
        });

        expect(updatedLinks).toHaveLength(3);
        expect(updatedLinks.map((it) => it.innerHTML)).not.toEqual([
          "link 1",
          "link 2",
          "link 3",
        ]);
      });
    });
  });
  describe("Test the appearance and the preview", () => {
    it("should display the appearance form when the button is clicked", async () => {
      await renderPopup();
      await openAppearance();
      const inputs = (await screen.findByRole("form")).querySelectorAll(
        "input"
      );

      expect(inputs).toHaveLength(4);
      expect(inputs[0]).toHaveProperty("name", "title");
      expect(inputs[1]).toHaveProperty("name", "headerBackgroundColor");
      expect(inputs[2]).toHaveProperty("name", "linkFontColor");
      expect(inputs[3]).toHaveProperty("name", "iconColor");
    });
    it("should change the title on the preview when a title is typed", async () => {
      await renderPopup();
      await openAppearance();
      const titleInput = await screen.findByRole("textbox");
      await userEvent.type(titleInput, "Title");
      const preview = await screen.findByText("Title");
      expect(preview).to.exist;
    });
    it("should change the background color of the header if the color is changed", async () => {
      await renderPopup();
      await openAppearance();
      const previewHeader = (
        await screen.findByTestId("preview")
      ).querySelector(".preview__card--header");
      expect(getComputedStyle(previewHeader).backgroundColor).toBe(
        "rgb(248, 249, 248)"
      );
      const form = await screen.findByRole("form");
      const headerColorInput = form.querySelector("input#header-bg");
      const headerColorLabel = form.querySelector("span.header");
      await act(async () => {
        await userEvent.pointer(headerColorInput, {});
        fireEvent.change(headerColorInput, {
          target: { value: "#f2f2f2" },
        });
      });
      expect(headerColorLabel.innerHTML).toBe("#f2f2f2");
      expect(getComputedStyle(previewHeader).backgroundColor).toBe(
        "rgb(242, 242, 242)"
      );
    });
    it("should change the link text color if the color is changed", async () => {
      await renderPopup();
      await addNewLink("link 1", "http://localhost:4321");
      await openAppearance();
      const previewLink = (await screen.findByTestId("preview")).querySelector(
        "li.preview__card--item a"
      );
      expect(getComputedStyle(previewLink).color).toBe("rgb(52, 64, 84)");
      const form = await screen.findByRole("form");
      const linkColorInput = form.querySelector("input#link-color");
      const linkColorLabel = form.querySelector("span.link");
      await act(async () => {
        fireEvent.change(linkColorInput, {
          target: { value: "#000" },
        });
      });
      expect(linkColorLabel.innerHTML).toBe("#000000");
      expect(getComputedStyle(previewLink).color).toBe("rgb(0, 0, 0)");
    });
    it("should change the icon color if the color is changed", async () => {
      await renderPopup();
      await addNewLink("link 1", "http://localhost:4321");
      await openAppearance();
      const previewIcon = (await screen.findByTestId("preview")).querySelector(
        ".preview__card--icon"
      );
      expect(previewIcon.querySelector("path").getAttribute("stroke")).toBe(
        "#7F56D9"
      );
      const form = await screen.findByRole("form");
      const iconColorInput = form.querySelector("input#icon");
      const iconColorLabel = form.querySelector("span.icon");
      await act(async () => {
        fireEvent.change(iconColorInput, {
          target: { value: "#000" },
        });
      });
      expect(iconColorLabel.innerHTML).toBe("#000000");
      expect(previewIcon.querySelector("path").getAttribute("stroke")).toBe(
        "#000000"
      );
    });
  });
});
