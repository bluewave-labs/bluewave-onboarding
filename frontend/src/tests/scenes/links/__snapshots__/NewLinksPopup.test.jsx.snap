// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Helper Link popup > Tests if the popup is rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="new__container"
  />
</DocumentFragment>
`;
